### PR TITLE
fix: modify preprocessing to filter reads from chroms not in ref fasta

### LINKED ISSUE #86: reads_to_bigwig fails when fragments contain reads from chromosomes not in fasta
**Description**
When running `reads_to_bigwig.py`, fragments with reads not in the reference fasta cause the process to fail.

```
Traceback (most recent call last):
  File "/software/miniconda3/envs/r-chrombpnet2/lib/python3.8/site-packages/pyfaidx/__init__.py", line 1027, in __getitem__
    return self.records[rname]
KeyError: 'KI270713.1'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "dir/chrombpnet/chrombpnet/helpers/preprocessing/reads_to_bigwig.py", line 100, in <module>
    main(args)
  File "dir/chrombpnet/chrombpnet/helpers/preprocessing/reads_to_bigwig.py", line 70, in main
    plus_shift, minus_shift = auto_shift_detect.compute_shift(args.input_bam_file,
  File "/software/miniconda3/envs/r-chrombpnet2/lib/python3.8/site-packages/chrombpnet/helpers/preprocessing/auto_shift_detect.py", line 198, in compute_shift
    plus_pwm, minus_pwm = get_pwms(sampled_plus_reads, sampled_minus_reads, genome_fasta_path)
  File "/software/miniconda3/envs/r-chrombpnet2/lib/python3.8/site-packages/chrombpnet/helpers/preprocessing/auto_shift_detect.py", line 95, in get_pwms
    cur = str(g[x['chr']][int(x['start'])-20:int(x['start'])+20])
  File "/software/miniconda3/envs/r-chrombpnet2/lib/python3.8/site-packages/pyfaidx/__init__.py", line 1029, in __getitem__
    raise KeyError("{0} not in {1}.".format(rname, self.filename))
KeyError: 'KI270713.1 not in dir/refs/hg38/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta.'
```

**Desired Behavior**
Have a warning for genes not found in reference chrom regions, but still continue to output for genes that do match. This is helpful for if data is from collaborators and we only care about parts that match, even if the data may contain small outliers.

Alternatively, maybe keep error as the default behavior, but have an option to do the above?

### PR DESCRIPTION
**Pull request by riyavsinha**
Fixes #86

**Example Output**
Screen Shot 2023-02-22 at 8 42 57 PM

### CODE REVIEW

**Comment by riyavsinha** [comment]
Also right now the filtering stage is happening twice, once in `sample_reads` and once again in `generate_bigwig` which is a bit of a performance hit I think. However, deduping that will require some form of passing around the tempfile so callers can create streams of it instead of the raw input. Wanted to confirm that was fine before doing so / get thoughts on the best way to do that!

**Comment by suragnair** [comment]
Performance requirements are minimal since it's a one time step. Also it's reading the file twice in any case, so filtering is not a big overhead. Other steps (e.g. sorting the file before making bigwig) are slower.

**Comment by suragnair** [comment]
Can you run it on a couple of your files and check how much longer the new version (with filtering) take vs the previous? And also ensure that the outputs are identical.

**Comment by panushri25** [comment]
@suragnair and @riyavsinha thank you for the PR! Let me know when the code is ready for merge, I will also review and test run it from my end. 

If you have already tested the script yourself, please add a note in the PR. If you can provide an example data file (where this error occurred and is fixed now) that will be helpful for me to test run it! Thank you again!!

File: chrombpnet/helpers/preprocessing/auto_shift_detect.py [hunk 1]
```diff
@@ -35,6 +38,22 @@ def convolve(to_scan, longer_seq):
     for i in range(len(longer_seq) - len(to_scan) + 1):
         vals.append(np.sum(to_scan*longer_seq[i:i+len(to_scan)]))
     return vals
+    
+def filtered_tagalign_stream(tagaligns_stream, genome_file):
```
Comment (suragnair) [inline chrombpnet/helpers/preprocessing/auto_shift_detect.py]: can you add a one liner describing the function here as well

**Comment by riyavsinha** [comment]
Ah good idea to do the comparison, below are the results. Output is the same also between running the modified pipeline on unfiltered reads, and unmodified on filtered reads, checked via `diff` which gave no output on the 3 test files.

Size | Filtered | Unfiltered
--- | --- | ---
285M | real 1m36.576s, user 1m24.658s, sys 0m39.292s | real 2m18.996s, user 1m35.426s, sys 0m42.820s
4.2G | real 17m1.127s, user 17m14.451s, sys 3m31.371s | real 20m31.623s, user 19m57.538s, sys 4m9.764s
10G | real	39m58.416s, user 41m36.133s, sys 8m0.258s | real 47m44.748s, user 47m23.883s, sys 8m57.656s

You can find the files in `/users/riyasinh/projects/aav-human/chrombpnet_timetest` if you would like to run it yourself on the data as well

**Comment by suragnair** [comment]
Looks like unfiltered is taking longer? That may be cuz it uses more reads and other steps are slower. For a fair comparison you can pre-filter the file and then run both versions.

File: chrombpnet/helpers/preprocessing/auto_shift_detect.py (hunk 1 above)
Comment (riyavsinha) [reply chrombpnet/helpers/preprocessing/auto_shift_detect.py]: Done

File: chrombpnet/helpers/preprocessing/reads_to_bigwig.py [hunk 2]
```diff
@@ -36,14 +36,14 @@ def generate_bigwig(input_bam_file, input_fragment_file, input_tagalign_file, ou
...
 
     cmd = """awk -v OFS="\\t" '{{if ($6=="+"){{print $1,$2{0:+},$3,$4,$5,$6}} else if ($6=="-") {{print $1,$2,$3{1:+},$4,$5,$6}}}}' | sort -k1,1 | bedtools genomecov -bg -5 -i stdin -g {2} | bedtools sort -i stdin """.format(plus_shift_delta, minus_shift_delta, chrom_sizes_file)
 
     tmp_bedgraph = tempfile.NamedTemporaryFile()
     print("Making BedGraph")
     with open(tmp_bedgraph.name, 'w') as f:
-        p2 = subprocess.Popen([cmd], stdin=p1.stdout, stdout=f, shell=True)
+        p2 = subprocess.Popen([cmd], stdin=subprocess.PIPE, stdout=f, shell=True)
+        auto_shift_detect.stream_filtered_tagaligns(p1, genome_fasta_file, p2)
         p1.stdout.close()
```
Comment (suragnair) [inline chrombpnet/helpers/preprocessing/reads_to_bigwig.py]: I think this can be removed?

Comment (suragnair) [reply chrombpnet/helpers/preprocessing/reads_to_bigwig.py]: as it's closed within `auto_shift_detect.stream_filtered_tagaligns`

Comment (riyavsinha) [reply chrombpnet/helpers/preprocessing/reads_to_bigwig.py]: Good catch! Done

**Comment by suragnair** [comment]
great can you benchmark and test one more time, thanks!

**Comment by riyavsinha** [comment]
Benchmark run on 3 files of different sizes on the master branch (unmodified current version) and the modified branch (incoming version).

Tagalign Size | Runtime (current version) | Runtime (incoming version)
--- | --- | ---
285M | real 1m36.576s, user 1m24.658s, sys 0m39.292s | real 1m38.999s, user 1m34.078s, sys 0m38.734s
4.2G | real 17m1.127s, user 17m14.451s, sys 3m31.371s | real 18m1.143s, user 19m59.964s, sys 4m4.964s
10G | real	39m58.416s, user 41m36.133s, sys 8m0.258s | real 42m16.157s, user 47m48.266s, sys 8m30.644s

No difference between the modified and unmodified outputs as well:
Screen Shot 2023-02-23 at 2 11 07 PM

**Comment by riyavsinha** [comment]
@panushri25 I think its ready for review :)

**Review by suragnair — APPROVED** [review]

**Comment by suragnair** [comment]
@riyavsinha can you change function name to something like `filter_tagalign_chromosomes` cuz it's not returning a stream now?

**Comment by riyavsinha** [comment]
I was thinking that while its not returning a stream, it is "streaming" as a verb data from p1 to p2, and so it might make it clear that that was how the data was getting between the two since right now p1 just outputs to a pipe and p2 takes in a pipe? If `filter_tagalign_chromosomes` still makes more sense to you though happy to change it

**Comment by suragnair** [comment]
ok sounds good to me! All clear to merge.

**Review by panushri25 — APPROVED** [review]

File: chrombpnet/helpers/preprocessing/auto_shift_detect.py [hunk 3]
```diff
@@ -35,6 +37,27 @@ def convolve(to_scan, longer_seq):
...
+    Returns:
+        Boolean. Indicates whether any reads not in the reference fasta were 
+        detected.
+    '''
+    has_unknown_chroms = False
+    with pyfaidx.Fasta(genome_file) as g:
+        for line in iter(src_tagaligns_stream.stdout.readline, b''):
+            tagalign_chrom = line.decode('utf-8').strip().split('\t')[0]
+            if tagalign_chrom in g.keys():
+                out_stream.stdin.write(line)
```
Comment (panushri25) [inline chrombpnet/helpers/preprocessing/auto_shift_detect.py:56]: @riyavsinha is this writing to the `/tmp/` directory?

Comment (panushri25) [reply chrombpnet/helpers/preprocessing/auto_shift_detect.py:56]: A user reported an error at this line ... `BrokenPipeError: [Errno 32] Broken pipe` and `sort: write failed: /tmp/sortU7nWxt: No space left on device`

Comment (panushri25) [reply chrombpnet/helpers/preprocessing/auto_shift_detect.py:56]: Exact command they used if helpful - this is on the tutorial data and run on Sherlock 

`sbatch -p gpu,owners --gpus 1 -c 10 --mem 120GB --time 1-0 --job-name=${job_header} --output=${job_header}.out --error=${job_header}.err --wrap singularity exec --contain --nv -e --no-mount hostfs /oak/stanford/groups/wjg/skim/software/sing_containers/chrombpnet.sif chrombpnet pipeline -ibam ${bam} -d ${mode} -g ${genome} -c ${chromsize} -p ${peaks} -n ${nonpeaks} -fl ${fold} -b ${biasmodel} -o ${outfile}`

**Comment by panushri25** [comment]
@suragnair there is a issue with this PR, let me know if you have any thoughts. Riya said she wont have time to look at it for now and is okay with me reverting back on these changes.

File: chrombpnet/helpers/preprocessing/auto_shift_detect.py (hunk 3 above)
Comment (suragnair) [reply chrombpnet/helpers/preprocessing/auto_shift_detect.py:56]: `No space left on device` likely means they don't have disk space?

Comment (panushri25) [reply chrombpnet/helpers/preprocessing/auto_shift_detect.py:56]: They have sufficient space on the disk they are running it on. Wondering why this command writes to the `/tmp/` folder and how much we can write to this folder on Sherlock.

Comment (suragnair) [reply chrombpnet/helpers/preprocessing/auto_shift_detect.py:56]: Error seems to be with sorting though. The file `/tmp/sortU7nWxt` seems to be a unix temp file.

Comment (suragnair) [reply chrombpnet/helpers/preprocessing/auto_shift_detect.py:56]: Changing the `TMPDIR` environment variable may be an option.

Comment (suragnair) [reply chrombpnet/helpers/preprocessing/auto_shift_detect.py:56]: This specific PR does not increase file or memory usage as it only introduces an extra filtering step which works with streams.

**Pull request merged**